### PR TITLE
Act on compass read return value

### PR DESCRIPTION
**Pull request by martinbudden**
Stops ignoring the return value of the compass read function. Takes action if read returns false.

### CODE REVIEW

**Comment by stronnag** [comment]
Very interesting. I've just rebuilt the quad with the dodo to eliminate the GPS/compass, so it'll probably be Monday before I try this.

**Comment by digitalentity** [comment]
Good catch!

HMC5883 is extremely picky. A few corrupt readings in self calibration and
compass will be totally off.

Next thing would be signalling this to IMU and navigation to take
appropriate action, but that's a task for another PR.

File: src/main/sensors/compass.c [hunk 1]
```diff
@@ -73,15 +73,18 @@ void updateCompass(flightDynamicsTrims_t *magZero)
 {
     static uint32_t calStartedAt = 0;
     static int16_t magPrev[XYZ_AXIS_COUNT];
-    uint32_t axis;
 
-    mag.read(magADCRaw);
-    for (axis = 0; axis < XYZ_AXIS_COUNT; axis++) magADC[axis] = magADCRaw[axis];  // int32_t copy to work with
+    if (!mag.read(magADCRaw)) {
+        return;
```
Comment (digitalentity) [inline src/main/sensors/compass.c:81]: We still need to return zero vector here so IMU can detect invalid compass and act accordingly.

**Comment by martinbudden** [comment]
I've updated to return a zero value in `sensors/compass.c` if the `mag.read` fails.

One other thought: in initialisation it tries to collect 10 samples, if any read fails there will be less than ten samples. Should I change it to collect 10 valid reads? And put in some protection so that if the reads continually fail it does not get stuck in the loop?

**Comment by digitalentity** [comment]
Yes, I think we should collect ten valid readings and if it times out just
return false.

Or maybe (for safety reasons) if we got at least one failed read/write
during detection process - report no compass.

**Comment by martinbudden** [comment]
OK, I've improved the handling of invalid readings in initialisation.

The `init` function is `void` not `bool` - I think that should be changed as well, but that is more pervasive, so I'll wait until @stronnag has tested this before I make that change.

**Comment by digitalentity** [comment]
I agree to both:
- we need to test and merge this first
- init function should be able to fail as well and the result of failure should be a non-present sensor indication.

**Comment by digitalentity** [comment]
I think it's safe to merge this right away. @martinbudden is this ready to merge?

**Comment by digitalentity** [comment]
See #514 for further discussion.

**Comment by martinbudden** [comment]
Yes, you can merge this now.

**Pull request merged**